File: src/data/types/hireResult.ts
import type { CrewMember } from "./crewMember.ts";
import type { PlanetCrewPool } from "./planetCrewPool.ts";
import type { Wallet } from "./wallet.ts";

// Necessary completion: Agent 16's contract names `hireCrew(...): CrewMember
// | HireError` but never defines HireError, and a bare CrewMember return
// on success can't also carry the updated pool/wallet a pure function
// must return alongside it. Modeled on the existing CraftResult/
// PurchaseResult pattern (a discriminated union), since capacity-full and
// insufficient-funds are normal business outcomes the caller must always
// handle, not exceptional cases.
export interface HireSucceeded {
  hired: true;
  crewMember: CrewMember;
  updatedPool: PlanetCrewPool;
  updatedWallet: Wallet;
}

export interface HireRejected {
  hired: false;
  reason: string;
}

export type HireResult = HireSucceeded | HireRejected;

File: unity/ProfitableCore/Schema/HireResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/hireResult.ts's HireResult = HireSucceeded |
// HireRejected discriminated union, same sealed-class-hierarchy idiom as
// CraftResult/PurchaseResult.
public abstract class HireResult
{
    public abstract bool Hired { get; }
}

public sealed class HireSucceeded : HireResult
{
    public override bool Hired => true;
    public CrewMember CrewMember { get; init; } = new();
    public PlanetCrewPool UpdatedPool { get; init; } = new();
    public Wallet UpdatedWallet { get; init; } = new();
}

public sealed class HireRejected : HireResult
{
    public override bool Hired => false;
    public string Reason { get; init; } = string.Empty;
}
